File: 48-rotate-image/rotate-image.py
class Solution:
    def transpose(self, matrix: List[List[int]]):
        for i in range(len(matrix[0])):
            for j in range(i+1,len(matrix[0])):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
    
    def reverse(self, matrix: List[List[int]]):
        
        for i in range(len(matrix)):
            end = len(matrix[0]) - 1
            front = 0
            while front < end:
                matrix[i][front],matrix[i][end] = matrix[i][end],matrix[i][front]
                front += 1
                end -= 1
            
                

            
            
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        self.transpose(matrix)
        self.reverse(matrix)

File: 48-rotate-image/rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length;
        transpose(matrix);
        reverse(matrix);
        
    }

    public void transpose(int[][] matrix) {
        int n = matrix.length; 

        for(int i=0; i<n; i++) {
            for(int j=i+1;  j<n; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }
    }

    public void reverse(int[][] matrix) {
        int n = matrix.length; 

        for(int i=0; i<n; i++) {
            int first = 0;
            int last = n-1;

            while (first < last) {
                int temp = matrix[i][first];
                matrix[i][first] = matrix[i][last];
                matrix[i][last] = temp;
                first++;
                last--;
            }

            
        }
    }
}
